Python/ibmcloudsql/SQLQuery.py

```diff
@@ -136,72 +211,296 @@ def _send_req(self, json_data):
 
             # Throw in case we hit the rate limit
             if (response.status_code == 429):
-                raise RateLimitedException("SQL submission failed: {}".format(response.json()['errors'][0]['message']))
+                time.sleep(3)  # seconds
+                raise RateLimitedException(
+                    "SQL submission failed ({code}): {msg}".format(
+                        code=response.status_code,
+                        msg=response.json()['errors'][0]['message']))
 
             # any other error but 429 will be raised here, like 403 etc
             response.raise_for_status()
 
             resp = response.json()
-            return resp['job_id']
-        except KeyError as e:
-            raise SyntaxError("SQL submission failed: {}".format(response.json()['errors'][0]['message']))
+            if 'job_id' in resp:
+                return resp['job_id']
+            else:
+                raise HTTPError()
+        except (KeyError, HTTPError) as _:
+            raise SyntaxError(
+                "SQL submission failed ({code}): {msg} - {query}".format(
+                    code=response.status_code,
+                    msg=response.json()['errors'][0]['message'],
+                    query=pformat(json_data)))
+
+    def submit(self,
+                   pagesize=None,
+                   blocking=True):
+        """ run the internal SQL statement that you created using the APIs provided by SQLMagic
+        """
+        self.format_()
+        return self.submit_sql(self._sql_stmt,
+            pagesize=pagesize,
+            blocking=blocking)
+
+    def submit_sql(self, sql_stmt, pagesize=None, blocking=True):
```

Comment:
Why introduce blocking and even make it default? I don't think this is correct semantics that you have in mind here. Setting max_tries alone is sufficient. No need for another parameter (and a value of 1000 max_tries that you set further below is way way too high to make sense anyway since_max tries will lead to exponentially increasing intervals between retries. 1000 retries is infinity). "blocking" anyway would rather suggest that the call is synchronous for SQL execution, which is not the case.

Reply:
The run_sql() is the version that wait until the result is done. 

Here, the submit_sql() is the version that waits until the job_id is available (which we consider as an asynchronous version). However, the job_id may not be available immediately if the capacity is reached. That's why the blocking mode comes in, i.e. keep sending the job until the job_id is returned. Even though max_tries can serve that purpose, its default value doesn't guarantee the job is launched (not completed) successfully. In my original implementation, 'blocking=True' means an infinite loop of pooling the job queue until it is submitted successfully. However, the use of the backoff package prevents me from doing this (at least I could not find a way to impose infinite loop), and I set to the very high values. Also, the interval between tries is not infinite, as I also impose a delay (3seconds) between tries.

Reply:
I really don't see a reason fort adding the blocking parameter. All can be controlled by max_tries parameter. As said, the blocking parameter just invokes confusion because the first assumption will be that it blocks until job has finished, not just submitted.

Reply:
Yup, I removed `blocking` option in these APIs as requested.